app/src/main/java/com/infomaniak/drive/data/services/BulkDownloadWorker.kt

```diff
@@ -0,0 +1,286 @@
+        numberOfFilesDownloaded += 1
+
+        val progressPercent = (numberOfFilesDownloaded * 100) / filesCount
+        val notificationContentTitle = applicationContext.getString(
+            R.string.bulkDownloadNotificationTitleWithProgress,
+            progressPercent
+        )
+        val notificationContentText = applicationContext.resources.getQuantityString(
+            R.plurals.bulkDownloadNotificationContent,
+            numberOfFilesDownloaded,
+            numberOfFilesDownloaded,
+            filesCount
+        )
+        updateDownloadNotification(
+            contentTitle = notificationContentTitle,
+            contentText = notificationContentText,
+            progressPercent = progressPercent
+        )
+    }
+
+    private fun notifyDownloadCancelled() {
+        Intent().apply {
+            action = TAG
+            LocalBroadcastManager.getInstance(applicationContext).sendBroadcast(this)
+        }
+    }
+
+    private fun saveRemoteData(
+        response: Response,
+        outputFile: IOFile? = null,
+        outputStream: AutoCloseOutputStream? = null,
+        onFinish: (() -> Unit)? = null
+    ) {
+        SentryLog.d(TAG, "save remote data to ${outputFile?.path}")
+        BufferedInputStream(response.body?.byteStream()).use { input ->
+            val stream = outputStream ?: outputFile?.outputStream()
+            stream?.use { output ->
+                input.copyTo(output)
+                onFinish?.invoke()
+            }
+        }
+    }
+
+    private fun downloadFileResponse(
+        fileUrl: String,
+        okHttpClient: OkHttpClient = HttpClient.okHttpClient,
+        onProgress: (progress: Int) -> Unit
+    ): Response {
+        val request = Request.Builder().url(fileUrl).headers(HttpUtils.getHeaders(contentType = null)).get().build()
+
+        return okHttpClient.newBuilder()
+            .addNetworkInterceptor(downloadProgressInterceptor(onProgress)).build()
+            .newCall(request).execute()
+    }
+
+    private fun downloadProgressInterceptor(onProgress: (progress: Int) -> Unit) = Interceptor { chain: Interceptor.Chain ->
+        val originalResponse = chain.proceed(chain.request())
+
+        originalResponse.newBuilder()
+            .body(ProgressResponseBody(originalResponse.body!!, object : ProgressResponseBody.ProgressListener {
+                override fun update(bytesRead: Long, contentLength: Long, done: Boolean) {
+                    val currentSystemTimeMillis = System.currentTimeMillis()
+                    if (currentSystemTimeMillis - lastUpdateProgressMillis > MAX_INTERVAL_BETWEEN_PROGRESS_UPDATE_MS) {
+                        lastUpdateProgressMillis = currentSystemTimeMillis
+                        val progress = (bytesRead.toFloat() / contentLength.toFloat() * 100F).toInt()
+                        onProgress(progress)
+                    }
+                }
+            })).build()
+    }
```

Comment:
Avoid duplication: when a function exists, use it whenever possible, otherwise create a utility, but avoid duplicating functions.

Reply:
I agree for `downloadFileResponse` but `downloadProgressInterceptor` is different because we do not want to update the progress too often. I'll move `downloadFileResponse` to `CoroutineWorkerExtensions`

Reply:
I understand, but you can still use the same method and just ignore the progression when it's not necessary since that's the only difference